### ISSUE TITLE
Ingest data from the CS&S AirTable

### ISSUE THREAD

**Issue by choldgraf** [2i2c-org/team-compass#545]
### Context

Code for Science and Society uses AirTable extensively for their bookkeeping and accounting. They use it to generate financial statements for 2i2c every month (which are placed in a Google drive).

We'd like to have access to that data in order to generate our own reports (e.g. at https://github.com/2i2c-org/kpis). We'd also like to see if we can more directly integrate with their AirTable system in our leads tracking system[^1].

[^1]: - Ref issue: https://github.com/2i2c-org/leads/issues/94

### Proposal

We should meet with CS&S to understand two things:

- How can we ingest their AirTable data directly (via a token, authenticated access, etc)
- What structure does their AirTable data have, and how can we use it for ourselves
- How could we integrate our leads tracking with theirs

### Checkpoints

- [x] Reach out to CS&S to plan a meeting
- [x] Have the meeting and decide on next steps from there

**Comment by jmunroe** [comment]
I created a new (private) repo to explore building a dashboard: https://github.com/2i2c-org/dashboard

Within that repo, there is a [Jupyter notebook](https://github.com/2i2c-org/dashboard/blob/main/AirTableIntegration.ipynb) that gives some working examples of extracting data from AirTable into Pandas dataframe.

**Comment by damianavila** [comment]
Nice!! Wondering if there is info on AirTable about what type of hub we agreed on.
At the time to deploy a new hub having a quick glance at the SOW agreed type of hub might be useful for Eng to understand the historical context (and what we should deploy 😉).
It could be also useful at the time to revisit what we have deployed against what we agreed on.

**Comment by choldgraf** [comment]
Agreed that we should facilitate access for anybody to see the sow, because that is the "source is truth" for what we have committed to on a contract

**Comment by jmunroe** [comment]
Currently waiting on a response from CS&S for a meeting date.

**Comment by choldgraf** [comment]
### Update: now ingesting accounting data from CS&S via AirTable

I followed the great instructions that @jmunroe shared, and re-worked the KPIs visualization code in https://github.com/2i2c-org/kpis so that it uses AirTable API keys and doesn't require us to put any data in that repository directly.

Here's what I did:

- Created an AirTable API key for myself
- Created a new AirTable "Base" called Accounting, and I've added two tables to it.
  1. `Invoices` is the synced data from CS&S's invoicing table.
  2. `All Transactions` is a data-dump from the latest CS&S accounting report.
- I re-worked the kpis code so that it uses the AirTable API key and does all of the data ingestion/munging under-the-hood, so no data (other than the plots) is accessible unless you have the airtable API key

Will post an update to `kpis` in the near future but just wanted to share this as an FYI.

**Comment by choldgraf** [comment]
### Update: meeting with CS&S

An update from @jmunroe that he shared via slack:

> I've finished meeting with Page Metcalf and Joe Hand.  They will provide us with an AirTable view with all of the 2i2c-related 'contracts' information I need (it is a 2i2c-filtered view from their CS&S Main Database).  I think I am now also pretty clear on the use of "authorized" vs "paid", and how we should be understanding deferred revenue. It was a good meeting and I have no more outstanding questions for CS&S.
> One stream of data that is not easily available through AirTable is any credit card charges that are incurred by 2i2c. This data is all in Ramp and exported to Xero, but not synced with AirTable. (In Xero, that is transaction-level information which is something Xero does not allow API level access for).  Joe Hand will try and see if there is a way of importing that information to AirTable but I agreed with him that it is not a priority request.

@jmunroe could you either:

1. Make a PR to the Team Compass to share this knowledge with the rest of the team (e.g. link to the new airtable when we get it, definition of authorized vs. paid, and any other important information for understanding how this process works)
2. Update it here if it's not yet clear enough what exactly needs to be added to the team compass?

**Comment by choldgraf** [comment]
### We have access to invoices and contracts from CS&S

CS&S have given us a "read view" of two tables, one for invoices and another for contracts. I've synced each of these to the 2i2c workspace so that they'll update automatically as new records come in. I've also documented them in this PR:

- #583

### Thoughts on these airtables

For future reference, here are some things I've noticed:

- The `Contracts` AirTable doesn't have a field for "Institution" (AKA, what institution is 2i2c contracting with). This would be useful for us to quickly sort contracts by institutional ID. I think that there should be two levels here: **Institution** (e.g., `Columbia University`), **Group** (e.g. `LEAP` or `Pangeo`)
- The `Invoices` AirTable doesn't cross-link to contracts (at least in our view). IMO, each invoice should be linked to a contract.
- The `Key` seems like it might be automatically generated from some of the data above that seems to be missing (e.g. it seems like it has a structure like `{2i2c identifier}: {date/status} - {institution} {name}`.
